### ISSUE TITLE
Investigate oclapi server failures

### ISSUE THREAD

**Issue by rkorytkowski** [OpenConceptLab/DATIM-Metadata-Project#37]
The last failure was on 10/08/2017. The oclapi docker container crashed, whereas other services were still running. The ocl_solr service was unresponsive and seemed to be the reason why oclapi crached. Here's the relevant part of the oclapi logs:

```
[10/Aug/2017 06:01:03] "GET /concepts/ HTTP/1.0" 200 5402
[2017/08/10 06:01:13] INFO    : 1433.19932318 sec GET /orgs/CIEL/sources/CIEL/HEAD/mappings/?limit=25&verbose=true&page=6498 : 200
[10/Aug/2017 06:01:14] "GET /orgs/CIEL/sources/CIEL/HEAD/mappings/?limit=25&verbose=true&page=6498 HTTP/1.1" 200 8091
[2017/08/10 06:01:14] INFO    : root 172.18.0.1 GET /orgs/CIEL/sources/CIEL/versions/?limit=0&verbose=true&page=1
[2017/08/10 06:01:17] INFO    : 2.11378502846 sec GET /orgs/CIEL/sources/CIEL/versions/?limit=0&verbose=true&page=1 : 200
[10/Aug/2017 06:01:17] "GET /orgs/CIEL/sources/CIEL/versions/?limit=0&verbose=true&page=1 HTTP/1.1" 200 1200
[2017/08/10 06:02:36] INFO    :  45.55.50.162 GET /concepts/
Failed to query Solr using '(source_version:(5821b7a564d700001440f44b) AND retired:(false) AND is_active:(true))': Connection to server 'http://solr.openconceptlab.org:8983/solr/collection1/select/?q=%28source_version%3A%285821b7a564d700001440f44b%29+AND+retired%3A%28false%29+AND+is_active%3A%28true%29%29&fq=django_ct%3A%28mappings.mappingversion%29&rows=25&facet.field=fromConceptSource_exact&facet.field=fromConceptOwner_exact&facet.field=conceptSource_exact&facet.field=ownerType_exact&facet.field=conceptOwner_exact&facet.field=retired_exact&facet.field=collection&facet.field=toConceptSource_exact&facet.field=conceptOwnerType_exact&facet.field=toConceptOwnerType_exact&facet.field=fromConceptOwnerType_exact&facet.field=source_exact&facet.field=mapType_exact&facet.field=owner_exact&facet.field=toConceptOwner_exact&wt=json&start=137800&fl=%2A+score&facet=on' timed out: HTTPConnectionPool(host='solr.openconceptlab.org', port=8983): Read timed out. (read timeout=10)
[2017/08/10 06:04:32] INFO    :  45.55.50.162 GET /concepts/
[2017/08/10 06:04:32] INFO    :  45.55.50.162 GET /concepts/
[2017/08/10 06:04:33] INFO    :  45.55.50.162 GET /concepts/
[2017/08/10 06:04:38] INFO    :  45.55.50.162 GET /concepts/
[2017/08/10 06:06:43] INFO    : root 172.18.0.1 GET /orgs/CIEL/sources/CIEL/concepts/160831/?includeInverseMappings=true&verbose=true&includeMappings=true
[2017/08/10 06:09:00] INFO    :  45.55.50.162 GET /concepts/
[2017/08/10 06:09:00] INFO    :  45.55.50.162 GET /concepts/
[2017/08/10 06:09:01] INFO    :  45.55.50.162 GET /concepts/
[2017/08/10 06:09:01] INFO    :  45.55.50.162 GET /concepts/
[2017/08/10 06:09:01] INFO    :  45.55.50.162 GET /concepts/
[2017/08/10 06:09:01] INFO    :  45.55.50.162 GET /concepts/
[2017/08/10 06:09:01] INFO    :  45.55.50.162 GET /concepts/
[2017/08/10 06:09:01] INFO    :  45.55.50.162 GET /concepts/
startup.sh: line 25:   115 Killed                  python manage.py runserver 0.0.0.0:8000 --settings="oclapi.settings.$SETTINGS" --configuration="$CONFIG"
```

Unfortunately I didn't save ocl_solr logs properly as they were not included in docker-compose logs for some reason... The next time the server crashes, logs need to be taken directly from solr server with:

`docker cp ocl_solr:/opt/solr/server/logs/solr.log .`

**Comment by rkorytkowski** [comment]
I also notice that someone was doing the following rest request: https://api.openconceptlab.org/orgs/CIEL/sources/CIEL/HEAD/mappings/?limit=25&verbose=true&page=6498,
which takes over 2 minutes to complete and brings the server to its limits...  Paging must not be done right since there's a limit of 25 so it shouldn't be loading that much from the database...

**Comment by paynejd** [comment]
Interesting that solr logs are not included in docker-compose. Is that
something we should fix?

2 responses to the HEAD request that you found:

If you browse to https://www.openconceptlab.org/search/ it performs
whichever search you specifically requested (e.g. malaria), plus it
performs a HEAD request to each of the other resources with your search
criteria applied to populate the counts (e.g. 104, 0, 0, ...)

[image: image.png]
The GET request will hit Solr to determine which results to return, then
queries Mongo once per result to get the rest of the object attributes.

The HEAD request should only need to hit Solr, so it should be extremely
fast, but as I recall, the original developer implemented the HEAD request
by simply performing the full GET request but only returning the response
header. Given that the search page receives the majority of traffic on OCL,
fixing the HEAD requests (if they are in fact broken) could result in a big
performance improvement.

Finally, even if paging is done right, it's possible that the indexes in
Mongo aren't setup correctly, so pulling up those 25 mapping records means
25 independent mapping queries -- it's possible that's the issue.

**Comment by rkorytkowski** [comment]
I managed to bring the server down by running the heavy query 20 times or so.

I'm getting the logs now.

**Comment by rkorytkowski** [comment]
So this time solr and mongo are doing fine and the logs are correctly displayed in docker-compose logs... Nothing suspicious in solr logs. It responds in at most 2 seconds for queries (mostly under a second). Maybe I didn't ping solr correctly last time and it was running properly...

It's just the oclapi that crashed... The paging seems to be the issue... I'll continue that trail by opening a new issue... Thanks Jon for explaining the logic. It's best I'll dive into code.

**Comment by rkorytkowski** [comment]
I've added restarting the oclapi on failures, which should now happen automatically. @paynejd, feel free to push it to production, if you don't want to continue doing it manually until we fix #39.
